test/SILGen/function_conversion_objc.swift

```diff
@@ -113,3 +113,93 @@ func cFuncPtrConversionUnsupported(_ x: @escaping @convention(c) (@convention(bl
     -> @convention(c) (@convention(c) () -> ()) -> () {
   return x  // expected-error{{C function pointer signature '@convention(c) (@convention(block) () -> ()) -> ()' is not compatible with expected type '@convention(c) (@convention(c) () -> ()) -> ()'}}
 }
+
+// SR-12723
+func SR12723_thin(_: (@convention(thin) () -> Void) -> Void) {}
+func SR12723_block(_: (@convention(block) () -> Void) -> Void) {}
+func SR12723_c(_: (@convention(c) () -> Void) -> Void) {}
+
+func SR12723_function(_: () -> Void) {}
+
+func context() {
+  SR12723_c(SR12723_function) 
+  // expected-error@-1 {{function signature '(() -> Void) -> ()' is not compatible with expected type '(@convention(c) () -> Void) -> Void'}}
+  SR12723_block(SR12723_function) 
+  // expected-error@-1 {{function signature '(() -> Void) -> ()' is not compatible with expected type '(@convention(block) () -> Void) -> Void}}
+
+  // CHECK: function_ref @$s24function_conversion_objc12SR12723_thinyyyyyXfXEF : $@convention(thin) (@noescape @callee_guaranteed (@convention(thin) @noescape () -> ()) -> ()) -> ()
+  // CHECK: apply
+  SR12723_thin(SR12723_function) 
+}
+
+struct SR12723_C {
+  let function: (@convention(c) () -> Void) -> Void
+}
+
+struct SR12723_Thin {
+  let function: (@convention(thin) () -> Void) -> Void
+}
+
+struct SR12723_Block {
+  let function: (@convention(block) () -> Void) -> Void
+}
+
+func proxy(_ f: (() -> Void) -> Void) {
+  let a = 1
+  f { print(a) }
+}
+
+func cContext() {
```

Comment:
I'm not exactly sure with a matrix of conventions that should be valid to convert the parameter type, this validation was based on whether it was crashing reaching invalid states later down on SILGen. 
One example is in cases like 
```
let _ :  (() -> Void) -> Void = thin.function // thin -> swift
```
Hitting an invalid state 
https://github.com/apple/swift/blob/bfffbbf520b982a377a2150cf1e28c8326d2574f/lib/SILGen/SILGenExpr.cpp#L1550

TBH I don't understand much about SILGen and what kind of conversions are allowed/supported, so feedback is appreciated :)
cc @slavapestov @jckarter @gottesmm